manual/sphinx/developer_docs/data_types.rst

```diff
@@ -217,51 +217,201 @@ See the example ``examples/test-globalfield`` for more examples.
+parallelise and vectorise. Some tuning of this is possible, see below
+for details. It replaces the C-style triple-nested loop::
 
    Field3D f(0.0);
    for (int i = mesh->xstart; i < mesh->xend; ++i) {
      for (int j = mesh->ystart; j < mesh->yend; ++j) {
        for (int k = 0; k < mesh->LocalNz; ++k) {
-         f[i,j,k] = a[i,j,k] + b[i,j,k]
+         f(i,j,k) = a(i,j,k) + b(i,j,k)
        }
      }
    }
 
-The iterator provides access to the x, y, z indices::
+The region to iterate over can be over ``Field2D``, ``Field3D``, or
+``FieldPerp`` domains, obtained by calling functions on ``Mesh``:
+``getRegion2D("name")``, ``getRegion3D("name")`` and
+``getRegionPerp("name")`` respectively. Currently the available regions include:
+
+-  `RGN_ALL`, which is the whole mesh;
+
+-  `RGN_NOBNDRY`, which skips all boundaries and guard cells;
+
+-  `RGN_GUARDS`, which is only guard cells, both boundary and
+   communication cells;
+
+-  `RGN_NOX`, which skips the x boundaries and guard cells
+
+-  `RGN_NOY`, which skips the y boundaries and guard cells
+
+New regions can be created and modified, see section below.
+   
+A standard C++ range for loop can also be used, but this is unlikely
+to OpenMP parallelise or vectorise::
 
     Field3D f(0.0);
     for (auto i : f) {
-       f[i] = i.x + i.y + i.z;
+       f[i] = a[i] + b[i];
     }
 
-It is also possible to specify regions to iterate over using this
-syntax::
+If you wish to vectorise but can't use OpenMP then there is a serial
+verion of the macro:
+
+     BoutReal max=0.;
+     BOUT_FOR_SERIAL(i, region) {
+       max = f[i] > max ? f[i] : max;
+     }
+
+For loops inside parallel regions, there is ``BOUT_FOR_INNER``::
 
     Field3D f(0.0);
-    for (auto i : f.region(RGN_NOBNDRY)) {
-       f[i] = 1.0;
+    BOUT_OMP(parallel) {
+      BOUT_FOR_INNER(i, f.getMesh()->getRegion3D("RGN_ALL")) {
+         f[i] = a[i] + b[i];
+      }
+      ...
     }
+    
+If a more general OpenMP directive is needed, there is
+``BOUT_FOR_OMP``::
+
+  BoutReal result=0.;
+  BOUT_FOR_OMP(i, region, parallel for reduction(max:result)) {
+    result = f[i] > result ? f[i] : result;
+  }
+  
+The iterator provides access to the x, y, z indices::
 
-Available regions are:
+    Field3D f(0.0);
+    BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_ALL")) {
+      f[i] = i.x() + i.y() + i.z();
+    }
 
--  `RGN_ALL`, which is the whole mesh;
+Note that calculating these indices involves some overhead: The
+iterator uses a single index internally, so integer division and
+modulo operators are needed to calculate individual indices.
+
+To perform finite difference or similar operators, index offsets can
+be calculated::
+
+  Field3D f = ...;
+  Field3D g(0.0);
+  BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_NOBNDRY")) {
+    g[i] = f[i.xp()] - f[i.xm()];
+  }
+
+The ``xp()`` function produces an offset of ``+1`` in ``X``, ``xm()``
+an offset of ``-1`` in the ``X`` direction. ``xpp()``
```

Comment:
Actually these take an optional step size, which defaults to 1 (e.g. `xpp` is implemented as just `xp(2)`).